torchprime/torch_xla_models/model/base_causal_lm.py

```diff
@@ -0,0 +1,96 @@
+"""BaseCausalLM Module
+
+This module defines a minimal base class for causal language models using PyTorch.
+It includes a standard weight initialization method, a placeholder forward pass,
+and methods for saving and loading model checkpoints using the `safetensors` format.
+"""
+
+import json
+import os
+
+import torch
+import torch.nn as nn
+from huggingface_hub import snapshot_download
+from omegaconf import OmegaConf
+from safetensors import safe_open
+from safetensors.torch import save_file
+
+
+def load_sharded_safetensors_to_state_dict(model_dir: str) -> dict:
+  """Load a model state dict from sharded safetensors in a given directory."""
+  state_dict = {}
+  index_file = os.path.join(model_dir, "model.safetensors.index.json")
+  with open(index_file) as f:
+    index = json.load(f)
+  weight_map = index["weight_map"]
+  for filename in set(weight_map.values()):
+    path = os.path.join(model_dir, filename)
+    with safe_open(path, framework="pt", device="cpu") as f:
+      for key in f.keys():  # noqa: SIM118
+        state_dict[key] = f.get_tensor(key)
+  return state_dict
+
+
+def save_sharded_safetensors_by_layer(state_dict: dict, save_dir: str):
```

Comment:
if we want to consume produced checkpoint in HF, is there any other metadata or config that we should export or copy from original checkpoint?

How is it usually done?

Reply:
Model checkpoint is literately just a dict from tensor name to tensor, so I think there is no additional config apart from weight_map.

We also write our model config in the export method just for recording configs.